### DISCUSSION TITLE
Fee Markets for Parallel Transaction Execution

### DISCUSSION THREAD

**Discussion by aaronbuchwald** [ava-labs/hypersdk#1816] [Chat]
## Parallel Transaction Execution

Introducing parallel transaction execution while maintaining an upper bound on worst case performance introduces some complexity. If we remove every other bottleneck except transaction execution, we'll bottleneck on single core performance unless we can introduce some level of parallelism.

### Single Core Performance Limits

For a simple native transfer action [benchmark](https://github.com/ava-labs/hypersdk/blob/main/examples/morpheusvm/actions/transfer_test.go#L190), we can see that it takes ~1µ:

```
goos: darwin
goarch: arm64
pkg: github.com/ava-labs/hypersdk/examples/morpheusvm/actions
BenchmarkSimpleTransfer-12    	 1000000	      1011 ns/op	     216 B/op	       6 allocs/op
PASS
ok  	github.com/ava-labs/hypersdk/examples/morpheusvm/actions	3.550s
```

A naive EVM transfer action [benchmark](https://github.com/aaronbuchwald/hypersdk/blob/11f3b8922d9604a0e20c1ed9d81898020bdd73c0/examples/hyperevm/actions/evm_call_test.go#L586) is about 30x slower (note: there's likely room to optimize the shim layer here):

```
goos: darwin
goarch: arm64
pkg: github.com/ava-labs/hypersdk/examples/morpheusvm/actions
BenchmarkEVMTransfer-12    	  105064	     34275 ns/op	   15290 B/op	     166 allocs/op
PASS
ok  	github.com/ava-labs/hypersdk/examples/morpheusvm/actions	34.650s
```

This implies an upper bound of ~1m TPS (native action) or 30k TPS (general purpose/EVM) for single core execution (on my laptop).

## Naive Solution

A naive solution assuming there are no other bottlenecks is to simply allocate N cores and increase our throughput limit. But, this fails rather miserably if we look at the worst case performance.

To get a serialized execution (produce an equivalent result to serially executing transactions), transactions touching the same state must fall back to executing serially ie. if tx1 writes a value that tx2 reads, then tx2 has to wait for tx1 to finish writing that value. In other words, our worst case execution has not improved at all by introducing parallel execution.

The naive solution improves performance in the average case with some degree of parallelism, but does nothing to improve our worst case performance.

## Solution: Scheduler / Fee Market

To leverage parallel transaction execution, we need to introduce a scheduler and/or fee market that supports parallel execution and limits the maximum depth of dependent computation within a block. Our ideal solution should maintain an upper bound on worst case execution latency, while leveraging as much parallelism as the workload allows.

The simplest solution is probably to leave scheduling to the block builder and require that schedule maintains an upper bound on the expected computation time given a set number of cores.

Transactions declare the exact CPU units (could be expected microseconds of computation) and the block builder must produce a schedule that limits the expected execution time to 100ms.

When peers verify the block, they can either verify the schedule in advance or as they go and exit and mark the block as invalid if the schedule fails that requirement.

This design has the added advantage of leaving it to economically incentivized block builders to produce an efficient schedule and leaves room for scheduling improvements that do not necessarily require a hard fork.

Note: Solana does something similar, but their exact design and assumptions about whether the transactions have already replicated when they arrive at the scheduler are not fully clear to me from their design doc here: https://github.com/anza-xyz/agave/blob/master/docs/src/proposals/fee_transaction_priority.md.

**Comment by darioush** [comment]
One simple idea is to force the builder to make the block into N segments that are "conflict free", which can be run in parallel, presumably to reduce the total execution time of blocks (on average).
So valid blocks don't represent txs that read a piece of data in one segment and update it in another (but this still allows for 1/Nth of a block to update a single value, like minting nfts).

This means sometimes some blocks are more empty if all the transactions conflict. This approach only works with knowing state keys

**Reply by aaronbuchwald** [reply]
hmm I'm not sure exactly how we'd divide transactions into conflict free segments (certainly possible with additional restrictions, but not sure what it would look like).

Another idea (discussed offline as well) could be to split the block into two segments and dynamically allocate resources between them depending on demand:
1. Serial execution segment (could relax state key requirement)
2. Parallel execution (additional restrictions including state keys and potentially limited execution time)

The problem with serial execution is that it limits our ability to fully leverage hardware after reaching the point where tx execution becomes a bottleneck.

The problem with parallel execution is that it may require additional, complex restraints on transactions to define a scheduler/fee market that allows us to increase throughput while maintaining a tight limit on worst case execution.

If a chain requires both general purpose execution and wants to better leverage parallel execution for small, easy to schedule transactions a mixed model could work well.

For example, we can limit all transactions executed in parallel to take up the same small amount of CPU time to simplify the scheduling/pricing logic and scale execution of very light weight transactions.

If there's high demand for light weight transactions, the network can shift more resources to parallel execution and vice versa if there's greater demand for serial execution (longer transactions, possibly with access to the full state w/o state keys).

**Reply by aaronbuchwald** [reply]
Here's some (very) rough pseudo-code of what a scheduling algorithm would look like assuming:
1. Each transaction takes the same unit of time (simplifying assumption)
2. pre-defined transaction ordering
3. pre-defined number of cores

```golang
type Tx interface {
	StateKeys() []ids.ID
}

type TxHeap interface {
	HasNext() bool
	Push(txs []Tx)
	Next() (Tx, bool)
}

type Core interface {
	// Execute a transaction for a discrete quanta of time to simplify scheduling
	Execute(tx Tx)
}

func ScheduleTxs(txs TxHeap, cores []Core) {
	for {
		// Track conflicts within a given iteration over the available cores.
		// The first transcation will get the first core no matter what, and then
		// we'll take the next highest priority transaction that doesn't conflict
		// with anything we are executing in this iteration.
		txConflicts := set.NewSet[ids.ID](1)
		restoreTxs := make([]Tx, 0)
		for _, core := range cores {
			// Find the next highest priority transaction to execute and make sure
			// to mark conflicts and reschedule skipped transactions.
			selectedTx, skippedTxs, has := findNextTx(txs, txConflicts)
			if !has {
				break
			}
			txConflicts.Add(selectedTx.StateKeys()...)
			restoreTxs = append(restoreTxs, skippedTxs...)
			core.Execute(selectedTx)
		}
		// Push skipped transactions back to the heap and check if we're done.
		txs.Push(restoreTxs)
		if !txs.HasNext() {
			break
		}
	}
}

// findNextTx searches for the next transaction w/o conflicts in the heap
// returns the selected tx, a slice of transactions that do contain conflicts, and a success flag
func findNextTx(txs TxHeap, txConflicts set.Set[ids.ID]) (selectedTx Tx, restoreTxs []Tx, has bool) {
	restoreTxs = make([]Tx, 0)
	for {
		tx, ok := txs.Next()
		if !ok {
			return nil, restoreTxs, false
		}
		hasConflict := false
		for _, key := range tx.StateKeys() {
			if txConflicts.Contains(key) {
				hasConflict = true
				break
			}
		}
		if hasConflict {
			restoreTxs = append(restoreTxs, tx)
			continue
		}

		return tx, restoreTxs, true
	}
}
```

**Comment by najeal** [comment]
Based one the information and solana doc given upside, I started a draft of a scheduler (not integrated with chain.builder). The proposed way of keeping a theoric upper bound is done as following:
- a maximum CPU usage is declared (Umax)
- each core can use Umax as they are theorically processed in parallel
- each time a Tx is sent to a core, the core gauge is incremented by Tx CPU usage.
- when a waiting core (C1) receives a tx using previously blocked state keys, we burn available CPU usage of C1 until statekeys released time.

Example:
C1 has run 10 Tx and its current Uc1 has been increased to 100 units (Uc1 = 100 units)
Only blocked transactions are remaining -> C1 is waiting without doing any work.
C1 gets a Tx using a key which was blocking and has been released by C2 when Uc2 = 150 units.
50 units are burned for C1, we increase its Gauge to 150. Then we increase again the core gauge of the current Tx CPU units.
When all cores have reached the Umax, we stop adding Tx to the block.

Link to the poc: https://github.com/najeal/hypersdk/tree/scheduler-priority/internal/scheduler
